src/main/java/bio/terra/cli/command/resource/ListTree.java

```diff
@@ -0,0 +1,90 @@
+package bio.terra.cli.command.resource;
+
+import bio.terra.cli.businessobject.Context;
+import bio.terra.cli.businessobject.Resource;
+import bio.terra.cli.command.shared.BaseCommand;
+import bio.terra.cli.command.shared.options.WorkspaceOverride;
+import bio.terra.cli.serialization.userfacing.UFResource;
+import bio.terra.workspace.model.Folder;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import picocli.CommandLine;
+
+/** This class corresponds to the third-level "terra resource list-tree" command. */
+@CommandLine.Command(
+    name = "list-tree",
+    description = "List all resources and folders in tree view.")
+public class ListTree extends BaseCommand {
+  @CommandLine.Mixin WorkspaceOverride workspaceOption;
+
+  private final HashMap<UUID, ArrayList<UUID>> edges = new HashMap<>();
+  private final HashMap<UUID, String> idToName = new HashMap<>();
+  private final HashMap<UUID, Boolean> isFolder = new HashMap<>();
+  private final UUID root = UUID.randomUUID();
+  private final String TERRA_FOLDER_ID = "terra-folder-id";
+
+  /** List the resources and folders in the workspace. */
+  @Override
+  protected void execute() {
+    workspaceOption.overrideIfSpecified();
+
+    // Get all resources and folders in the workspace and sort by name
+    List<UFResource> resources =
+        Context.requireWorkspace().listResourcesAndSync().stream()
+            .sorted(Comparator.comparing(Resource::getName))
+            .map(Resource::serializeToCommand)
+            .collect(Collectors.toList());
+    List<Folder> folders =
+        Context.requireWorkspace().listFolders().stream()
+            .sorted(Comparator.comparing(Folder::getDisplayName))
+            .collect(Collectors.toList());
+
+    // Create edges map for DFS and store name for each id.
```

Comment:
Could 46-88 be simplified? I find this easier to understand:

```
  @Override
  protected void execute() {
    printFolder(/*folderId*/ 0, /*level=*/ 0);           // folderId=0 for root folder
  }

  printFolder(UUID folderId, int level) {
    listFoldersInFolder(folderId).forEach(
        folder -> {
          printLine(folder.getName(), level);
          printFolder(folder.getId(), level+1);
        }
    );

    listResourcesInFolder(folderId).forEach(
      resource -> printLine(resource.getName(), level);
    );
  }
```

Reply:
Hey Melissa, this is the iterative DFS I talked about and tried before. It works fine if we only need to print white space according to the level number. But later I found I would need another stack to store how many directory was visited in order to draw box-drawing character "└──" and "│   " each line. In this case, a recursion is much easier, and also the reason that was used in the GNU implementation.

Reply:
Interesting! Can you add comment on line 45, something like:

```
// Note: The intuitive algorithm doesn't handle drawing lines correctly, so use this algorithm from the
  // GNU Tree utility implementation: https://github.com/kddnewton/tree/blob/main/Tree.java
  // See https://github.com/DataBiosphere/terra-cli/pull/329/files#r982639897
```